metric_learn/base_metric.py

```diff
@@ -1,31 +1,76 @@
 from numpy.linalg import cholesky
 from sklearn.base import BaseEstimator
-from sklearn.utils.validation import check_array
+from sklearn.utils.validation import check_array, _is_arraylike
 from sklearn.metrics import roc_auc_score
 import numpy as np
 from abc import ABCMeta, abstractmethod
 import six
-from ._util import check_tuples
+from ._util import check_tuples, SimplePreprocessor
 
 
 class BaseMetricLearner(six.with_metaclass(ABCMeta, BaseEstimator)):
 
-  @abstractmethod
-  def score_pairs(self, pairs):
-    """Returns the score between pairs
-    (can be a similarity, or a distance/metric depending on the algorithm)
+  def __init__(self, preprocessor=None):
+    """
 
     Parameters
     ----------
-    pairs : `numpy.ndarray`, shape=(n_samples, 2, n_features)
-      3D array of pairs.
+    preprocessor : array-like, shape=(n_samples, n_features) or callable
+        The preprocessor to call to get tuples from indices. If array-like,
+        tuples will be gotten like this: X[indices].
+    """
+    self.preprocessor = preprocessor
+
+  @abstractmethod
+  def score_pairs(self, pairs):
+      """Returns the score between pairs
+      (can be a similarity, or a distance/metric depending on the algorithm)
+
+      Parameters
+      ----------
+      pairs : `numpy.ndarray`, shape=(n_samples, 2, n_features)
+        3D array of pairs.
+
+      Returns
+      -------
+      scores: `numpy.ndarray` of shape=(n_pairs,)
+        The score of every pair.
+      """
+
+  def check_preprocessor(self):
+    if _is_arraylike(self.preprocessor):
+      self.preprocessor_ = SimplePreprocessor(self.preprocessor)
+    else:
+      self.preprocessor_ = self.preprocessor
+
+  def metric(self):
+    """Computes the Mahalanobis matrix from the transformation matrix.
+
+    .. math:: M = L^{\\top} L
 
     Returns
     -------
-    scores: `numpy.ndarray` of shape=(n_pairs,)
-      The score of every pair.
+    M : (d x d) matrix
+    """
+    L = self.transformer()
+    return L.T.dot(L)
+
+  def transformer(self):
+    """Computes the transformation matrix from the Mahalanobis matrix.
+
+    L = cholesky(M).T
+
+    Returns
+    -------
+    L : upper triangular (d x d) matrix
     """
+    return cholesky(self.metric()).T
 
+  def format_input(self, input):
```

Comment:
choose more concrete name for `input` and the function name (preprocess instead of format?)